### ISSUE TITLE
Swap Delimiters Purges Undo Stack

### ISSUE THREAD

**Issue by AmeliaSZK** [DaxStudio/DaxStudio#255]
**Describe the bug**
After using the Swap Delimiters function, you cannot undo (Ctrl+Z) this action, nor any other previous actions. (DAX Studio 2.9.4)

**To Reproduce**
1. Create a new blank query. Click "Cancel" in the window to connect.
2. Write: `= CONCATENATE("one", " and two")`
3. Click on Swap Delimiters. You'll get: `= CONCATENATE("one"; " and two")`
4. Click on Undo, or use Ctrl+Z. It will not work.

**Expected behaviour**
The Undo should be able to reverse the Swap Delimiters action just as any other search-and-replace action.

**Additional context**
I use DAX Studio to format my formulas that I then copy-paste into Excel. But my Excel is in French, so I need to use the Swap Delimiters to make sure the copy-pasted formula has semicolons instead of commas. (Commas can sometimes work in Excel DAX formulas, but not always)

So the workflow is to Format Query, then Swap Delimiters, and then copy-paste.

**The problem comes with comments.** Format Query will silently remove comments that are placed at some wrong position. However, I've observed that Excel will *also* silently remove comments, so I'm assuming that DAX Studio removes those same comments. Which is fine, because if I see that a comment has been removed, I can Undo and then place it elsewhere, and try to find a safe and appropriate place by trial-and-error.  (I haven't found any documentation as to where comments will be silently removed, and I've searched on https://docs.microsoft.com/en-us/dax/ and googled "how to prevent excel from deleting dax comments". This seems to be a surprisingly niche issue?!?)

Anywayz. The **critical issue** is that if I don't notice that a comment was silently removed before doing the Swap Delimiters, well I will have permanently lost it. This lack of notice hasn't happened to me so far, but the potential is there with functions that don't fit on a single screen.

Comments are extra efforts someone put into their code, and losing them can be extremely disheartening. Furthermore, this permanent removal being caused by what feels like reasonable unspoken expectations can cause a complete loss of trust into the reliability of the software. Here, those specific assumptions are:
- Formatting tools are non-destructive to the content.
- Search-and-replace actions can be Undone.
- Swap Delimiters being a (somewhat elaborate) search-and-replace action is a legitimate user perception.

Feel free to let me know if any of these is actually not a reasonable assumption. I'm not a formally trained DAX user, so there might be some common field knowledge that I'm missing here.

**Additional expected behaviour**
Format Query should report an error if it finds a comment it would remove, just like when it tries to format formulas using the semicolon delimiter.

**TL;DR**
When using DAX Studio to format formulas for a locale with semicolon delimiters, comments are at risk of permanent loss because of this combination of circumstances:
- Swap Delimiters cannot be Undone.
- Format Query silently removes comment that are in some places. 

**Addendum (edit)**
I just remembered something: While I may have, in the beginning, perceived Swap Delimiters as *not* a search-and-replace, since replacing delimiters is *not* just replacing characters, I've came to view it as just a search-and-replace because it also permanently alters my punctuation in comments. So "then..." becomes "then,,,", periods become commas and stay commas, etc. etc. 

This isn't a problem I feel worthy of a formal issue for a software that is distributed for free and likely maintained by volunteers, but I feel like it's worth mentioning as to how a user can come to view the Swap feature as a search-and-replace?

Anywayz, feel free to ask if you want more explanations as why I'm saying it *permanently* alters my punctuation.

**Comment by dgosbell** [comment]
### Swap Delimiters

Thanks for reporting this, I had not noticed that the swap delimiters function cleared the undo stack. There was actually a simple fix for this and I've been able to correct this issue and the fix will go out as part of the next release. (I also fixed a couple of other features which had a similar issue).

It never hurts to raise an issue like this, sometimes (as in this case) it is a simple fix. :) 

Note that you can set a default separator style in the Options screen and this *should* work with the Format Query function but there was a minor bug with that too which I've also just fixed. 

### Format Query

The Format Query issue is a bit more involved. We don't actually do the formatting - we collaborated with the daxformatter.com website and we make a call to their site to do our formatting. They don't deliberately remove comments, but I know that the parser technology they are using has issues with trailing comments

eg.

```
EVALUATE
FILTER( mytable, -- this comment will remain
mytable[Column1] = "abc"  -- this comment will disappear
) 
```
will appear as follows after formatting:
```
EVALUATE
FILTER (
    mytable,
    -- this comment will remain
    mytable[Column1] = "abc"
)
```
Technically the parser is attaching the comment to the following DAX token, not to the token on the same line. It's not as neat as end of line comments, but one work around is to always put your comments on the line above the element you are commenting.

I'm not sure if the issue in Excel is the same or not, but daxformatter.com are not trying to duplicate this behaviour on purpose. They would like to fix this at some point, but it's a big job as it's probably going to require almost a complete re-write.

**Comment by SlayZeKyriarchy** [comment]
Oh wow, thanks for the quick fix and the guideline to protect my comments. 👀 Super appreciated! 😄

**Comment by AmeliaSZK** [comment]
@dgosbell 

In the [commit](https://github.com/DaxStudio/DaxStudio/commit/294f99fd3d147c7831b909ba9debb8f2fa779fec) to resolve this issue, I noticed a potential bug (daxformatter's fault, not yours) of critical severity (imo). This bug could alter the content of string literals during formatting, and it can be observed on the [daxformatter](https://www.daxformatter.com) website. 

### Is this bug actually introduced by the commit under discussion?
I don't know, and I will not be able to test this any time soon, because it would require me to build the latest developments of DAX Studio on my machine. 

I decided to quickly notify you, before I could test the presence of the bug, because I don't know when the next release will happen, and strings literals being alterable by the formatter felt like a reasonable report to fast-track. 

### Context
In the commit, [parameters](https://github.com/DaxStudio/DaxStudio/blob/294f99fd3d147c7831b909ba9debb8f2fa779fec/src/DaxStudio.UI/Model/DaxFormatter.cs#L194) are added to the call to daxformatter dot com when the European delimiters are selected in the option menu. (Lines 194 to 198 in link) 

However, when drafting another issue for this repository, I noticed that on the daxformatter website, if you request "Others (A; B; C / 1,234)" as separators, the content of string literals can be altered in at least one case. 

This website parameter is conceptually equivalent to what was introduced in the commit, which is why I think it is reasonable to suspect that this bug could also be present in the web API used by DAX Studio. 

### To reproduce
1. Go to https://www.daxformatter.com
2. Select "Others (A; B; C / 1,234)" separators
3. Format this formula:
```
=
IF (
    [Some Test];
    /* Hello, World 2.0! */
    "Hello, World 2.0!";
    /* Wow; Stupendous! */
    "Wow; Stupendous!"
)
```
On my phone, it returns this:
```
=
IF (
    [Some Test];
    /* Hello. World 2.0! */
    "Hello. World 2.0!";
    /* Wow, Stupendous! */
    "Wow, Stupendous!"
)
```
4. Notice that the comma and semicolon in the string literals were changed, respectively, to a period and comma.

### Expected Behaviour
Content of string literals should never be altered by a formatting tool.

### Miscellaneous
A bug report was sent to DAX Formatter. 

I am still new in using Github to collaborate, and am unsure wether this post should have been a dedicated "Issue" or a reply in this thread. Feel free to let me know if it should have been its own issue.

**Comment by marcosqlbi** [comment]
This is a known issue in DAX Formatter and not a DAX Studio one.

**Comment by marcosqlbi** [comment]
We have been able to make improvements to the service - but in DAX Studio they will be visible only in the next release.
